### PR TITLE
Add the RDM handler code.

### PR DESCRIPTION
**Pull request by nomis52**

### CODE REVIEW

File: firmware/src/rdm_handler.c [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+  // TODO(simon):
+  // We need to intercept calls to the SET_MODEL_ID pid, and use them to change
+  // the active model.
+
+  if (!g_rdm_handler.active_model) {
+    return;
+  }
+
+  int response_size = g_rdm_handler.active_model->request_fn(
+      header, param_data);
```
Comment (peternewman) [inline firmware/src/rdm_handler.c]: Will these not fit on the previous line, or one and align with the bracket?

File: firmware/src/rdm_handler.h [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+typedef void (*RDMHandlerSendCallback)(bool include_break,
+                                       const IOVec* data,
+                                       unsigned int iov_count);
+
+/**
+ * @brief The settings to use for the RDM Handler.
+ */
+typedef struct {
+  uint8_t uid[UID_LENGTH];  //!< The responder's UID.
+  uint16_t default_model;  //!< The default model to use
```
Comment (peternewman) [inline firmware/src/rdm_handler.h]: Should this be an enum throughout?

File: firmware/src/rdm_model.h [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+ * The deactivate function is called if this modle is active, and another model
+ * becomes active via a call to RDMHandler_SetActiveModel().
+ */
+typedef void (*RDMModelDeactivateFunction) ();
+
+/**
+ * @brief The function used to handle an RDM request.
+ * @param header The RDM request header.
+ * @param param_data The RDM parameter data, or NULL if there wasn't any.
+ * @returns The size of the RDM response, in g_responder.buffer. The sign is
```
Comment (peternewman) [inline firmware/src/rdm_model.h]: Use a pointer and another arg rather than overloading?

**Comment by peternewman** [comment]
Just a couple of comments

File: firmware/src/rdm_handler.c (hunk 1 above)
Comment (nomis52) [inline firmware/src/rdm_handler.c]: it missed by 1 character.

File: firmware/src/rdm_handler.h (hunk 2 above)
Comment (nomis52) [inline firmware/src/rdm_handler.h]: Probably, but I need to make sure it's extensible by others, possibly outside our codebase. I may change it later to an enum.

File: firmware/src/rdm_model.h (hunk 3 above)
Comment (nomis52) [inline firmware/src/rdm_model.h]: No, it gets too clumsy to pass around another pointer to every single PID handler.

**Comment by nomis52** [comment]
Alright, ready to go now.

File: firmware/src/rdm_handler.c (hunk 1 above)
Comment (peternewman) [inline firmware/src/rdm_handler.c]: Ah okay

File: firmware/src/rdm_handler.h (hunk 2 above)
Comment (peternewman) [inline firmware/src/rdm_handler.h]: Okay

File: firmware/src/rdm_model.h (hunk 3 above)
Comment (peternewman) [inline firmware/src/rdm_model.h]: Okay

File: firmware/src/rdm.h [hunk 4]
```diff
@@ -104,6 +103,11 @@ enum { RDM_DEFAULT_STRING_SIZE = 32 };
 enum { DUB_RESPONSE_LENGTH = 24 };
 
 /**
+ * @brief The size of an RDM frame.
+ */
+enum { RDM_MAX_FRAME_SIZE = 257 };
```
Comment (peternewman) [inline firmware/src/rdm.h]: Can we define this as the sum of its parts.

File: tests/Doxyfile [hunk 5]
```diff
@@ -688,7 +688,7 @@ RECURSIVE              = NO
 # Note that relative paths are relative to the directory from which doxygen is
 # run.
 
-EXCLUDE                = config config.h
+EXCLUDE                = config config.h mocks/
```
Comment (peternewman) [inline tests/Doxyfile]: I think you can drop the slash, config is a directory too isn't it?

File: firmware/src/rdm_handler.h [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+/**
+ * @brief Initialize the RDM Handler sub-system.
+ * @param settings The settings for this responder.
+ */
+void RDMHandler_Initialize(const RDMHandlerSettings *settings);
+
+/**
+ * @brief Add a model to the list of available models.
+ * @brief model_entry The ModelEntry to add.
+ * @returns true if the model was added, false if there wasn't enough slots or
```
Comment (peternewman) [inline firmware/src/rdm_handler.h]: SPaG weren't enough

**Comment by peternewman** [comment]
A few more comments.

File: firmware/src/rdm_handler.h (hunk 6 above)
Comment (nomis52) [inline firmware/src/rdm_handler.h]: Done

File: tests/Doxyfile (hunk 5 above)
Comment (nomis52) [inline tests/Doxyfile]: Done

File: firmware/src/rdm.h (hunk 4 above)
Comment (nomis52) [inline firmware/src/rdm.h]: Apparently not since it needs to be a compile time constant.

**Pull request merged**